### ISSUE TITLE
save shortcut right alt+s in article write

### ISSUE THREAD

**Issue by o10154016** [textpattern/textpattern#1385]
### Expected behaviour

in Polish language combination right alt+s should give letter "ś"

### Actual behaviour

now is impossible to use this letter "ś" because of save

### Steps to reproduce

pressing right alt + s on polish keyboard layout

#### Additional information

line 1156 in file textpattern.js:
} else if (key === 19 || ((e.metaKey || e.ctrlKey) && String.fromCharCode(key).toLowerCase() === 's'))

Textpattern version: 4.7.2

**Comment by bloatware** [comment]
Can not reproduce. Could you specify OS+browser, please?

**Comment by philwareham** [comment]
@bloke @bloatware I think this should fix the issue, yes?...

    } else if (key === 19 || ((e.metaKey || e.ctrlKey) && !(e.altKey) && String.fromCharCode(key).toLowerCase() === 's'))

Let me know, and I'll make the change.

**Comment by bloatware** [comment]
Ah yes, if you mean `ctrl+alt+s` then this should work.

**Comment by philwareham** [comment]
Doesn't that code I supplied *omit* the alt key? I thought it did. I vaguely remember some keyboard layouts map the ctrl and alt keys differently so this is a workaround for that scenario.

Dunno, I'm no JavaScript expert so I might be wrong here.

**Comment by Bloke** [comment]
Didn't know about this but [we're not alone](https://microsoftteams.uservoice.com/forums/555103-public/suggestions/32265802-alt-a-and-alt-s-affecting-polish-character-for-%C4%85) - even Microsoft Teams screwed up. And some folk say their key combos affect Romanian and German too.

The thing I can't fathom though is that some people mention Alt+s, others mention Ctrl+Alt+s, others mention Ctrl+s. I'm not sure which it is!

According to the [Polish virtual keyboard](http://polish.typeit.org/) if you hover over some of the lower row of letters, the keyboard shortcut tooltips show up as CTRL+S and CTRL+A. That means that other standard browser functionality like Select All wouldn't work. And before Txp hijacked Ctrl+S for Save, you wouldn't be able to use the shortcut to save a web page to your computer. And if you look at the second letter, it's allegedly Ctrl+C which means baked-in browser keyboard shortcuts for Copy would be broken too.

I'm confused!

**Comment by philwareham** [comment]
@o10154016 please can you try the amended line 1156 in your setup and let us know if that has any effect on the issue? Basically we need to ensure the alt+s is not hijacked by our JavaScript.

**Comment by bloatware** [comment]
> According to the Polish virtual keyboard if you hover over some of the lower row of letters, the keyboard shortcut tooltips show up as CTRL+S and CTRL+A.

Tooltips show `ALT+` for me, and `CTRL+A` works as usual. `CTRL+ALT+A` acts like `ALT+A`. And, according to [mozdocs](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/getModifierState), `AltGr` (right `ALT`) on Windows acts like `CTRL+ALT`. If so, @philwareham patch should help.

**Comment by Bloke** [comment]
> Tooltips show `ALT+` for me

Maybe they show as `CTRL+` for me because I'm on a Mac then.

**Comment by bloatware** [comment]
Then +1 confused. Should we introduce some `user_script` preference responsible for this and other client-side functionality? Dunno...

**Comment by Bloke** [comment]
Okay, just tested Txp using Presentation->Styles. On my Mac:

* `CMD+s` saves the Stylesheet.
* `CTRL+s` also saves the Stylesheet.
* `ALT+s` inserts scharfers s (`ß`) at the cursor position.

So if the OP is using a Mac and the keyboard shortcut for the diacritic mark is indeed `CTRL+s` then yes Txp will trample on this.

The question then is, how can we write JavaScript to test for `CTRL+s` = save on Windows (where the diacritic is entered via `ALT+s`) and test for `CMD+s` = save on Mac (where the diacritic is entered via `CTRL+s`). Hmmmm....

**Comment by Bloke** [comment]
[Might be useful](https://unixpapa.com/js/testkey.html) for testing keypresses.

Interestingly, while Mac's `CMD` is registered during keyup/down, it doesn't appear to be considered a modifier :\ It's registered as `metaKey` which can be detected (apparently) with `event.metaKey` so the solution might lie in testing, in pseudocode:

```
if (event.metaKey + S || (event.ctrlKey + S && is_not_mac)) {
   // perform save
}
```

Quite how `is_not_mac` is determined reliably I don't know.

**Comment by phiw13** [comment]
> So if the OP is using a Mac and the keyboard shortcut for the diacritic mark is indeed CTRL+s then yes Txp will trample on this.

If the OP is using a Mac and using the Polish keyboard, then he/she would use `ALT + s` (or Option + s) to input that keyboard. I tested a dev install with that keyboard (input source) active and the “save” keyboard shortcuts work just fine. `CTRL` is very rarely used on macOS for character input.

I can imagine on Windows being different.

*System Preferences > Keyboards > Input Sources – click the + at the bottom to add keyboards

**Comment by Bloke** [comment]
Okay, thanks. So maybe that Polish keyboard thing I linked to above is an anomaly when they allow `CTRL+s` to input diacritic characters. On my MacBook, though, if I use `Alt+s` on that Polish keyboard page, I get `ß` inserted in the textarea. If I use `CTRL+s` I get `ś`.

I'm using a standard British keymap on a US-style keyboard (which is strange because there's an `@` on the `2` key and that's what I get when I type it with shift. I'd expect with a British keymap that the character would render a double quote. Meh, not complaining).

**Comment by Bloke** [comment]
Further info, this time on Windows. In Txp:

* `CTRL+s` saves the Stylesheet.
* `WindowsKey+s` also saves the Stylesheet.
* `ALT+s` does nothing.
* `ALTGr+s` does nothing.

When running the testkey app linked above:

* Both Windows keys trigger `event.metaKey` so we should be safe permitting that on any platform.
* The 'popup menu windows key' triggers keyCode=93 but does not trigger `event.*Key` of any kind.
* `shift`, `ctrl` and `alt` trigger the `event.*Key` and keycodes you'd expect (shift=16, ctrl=17, alt=18).
* `Alt gr` sends keyCode 17 and 18 on both keydown and keyup, and only `event.ctrlKey` on keydown (17) i.e. it's roughly the same as regular `alt+ctrl` but without the `event.altKey`.

And in that Polish keyboard site on Windows (Chrome), the shortcut keys for inserting diacritics are `ALT+s`, `ALT+a`, `Alt+c` etc. The `Ctrl+` counterparts do regular browser shortcuts: save page as HTML, select all, copy, etc.

Does that give us any further info what "rules" we can employ to handle this more consistently?

**Comment by bloatware** [comment]
@Bloke what is your browser/version? The following seems to be true in my case (FF 64):

> At least as of Firefox 48, the ⊞ Windows key is not considered the "Meta" key. KeyboardEvent.metaKey is false when the ⊞ Windows is pressed.

**Comment by Bloke** [comment]
Cortana isn't enabled for me so that shortcut is defunct at the system level.

I'm using Win 7 and Chrome v71. You're right that in Firefox, the Windows key doesn't trigger `event.metaKey`. But that doesn't matter as it just means that `Win+s` in Chrome will Save, but won't in FF. Since we don't officially support `Win+s` anyway, it's of no consequence; allowing it is simply a byproduct of permitting `CMD+s` on Mac.

What's of more importance is *not* allowing `Ctrl+s` (on Mac, if possible) and *not* allowing `Alt+s` (on any platform) to trigger save. Of the two, I'd say the latter is more important but it depends if that Polish keyboard emulator is an accurate reflection of the keyboard mappings on MacOS at large.

**Comment by phiw13** [comment]
@Bloke 

I had a hard time trying too figure out how you managed to type that character (ś) with the CTRL key until I realised that you were testing on that virtual keyboard you linked to above - *and* you were using Firefox. With Safari, that site uses the option (`alt`) key (Chromium browsers also seem to get the CTRL configuration).

So - most important is preventing the `ALT+s` key combo to trigger save. `CTRL + s` is **not** a combo that would trigger a character input on macOS anyway, so there is no harm here.

**Comment by philwareham** [comment]
So, was there a final decision on this - it's kind of blocking release of 4.7.3? Cheers!

**Comment by bloatware** [comment]
I think what you suggest [here](https://github.com/textpattern/textpattern/issues/1385#issuecomment-458473301) should be added in any case, but seemingly only @o10154016 can test whether it fixes the issue.

**Comment by philwareham** [comment]
Patched in dev and 4.7.3 branches - please test. If I don't hear within 48 hours, I'll assume it's fixed and close issue. Cheers.

**Comment by petecooper** [comment]
Bumping this since 10 days has passed since https://github.com/textpattern/textpattern/issues/1385#issuecomment-463160308 -- @o10154016, have you been able to test the updated `textpattern.js` ([here](https://github.com/textpattern/textpattern/blob/dev/textpattern/textpattern.js) if you need a link)?
